File: examples/go/property-initializers/BoundedCounter_test.go
package contract

import (
	"testing"

	runar "github.com/icellan/runar/packages/runar-go"
)

// NOTE: Contracts with property initializers define an init() method that sets
// default values (e.g., Count = 0, Active = true). In native Go tests, you must
// call init() after constructing the struct to apply these defaults. At compile
// time, the Rúnar compiler bakes these defaults directly into the Bitcoin Script.

func TestBoundedCounter_DefaultInitializers(t *testing.T) {
	c := &BoundedCounter{MaxCount: 10}
	// Call init() to apply property initializer defaults
	c.init()
	if c.Count != 0 {
		t.Errorf("expected Count=0, got %d", c.Count)
	}
	if c.Active != true {
		t.Errorf("expected Active=true, got %v", c.Active)
	}
	c.Increment(1)
	if c.Count != 1 {
		t.Errorf("expected Count=1 after increment, got %d", c.Count)
	}
}

func TestBoundedCounter_Increment(t *testing.T) {
	c := &BoundedCounter{MaxCount: 10}
	c.init()
	c.Increment(3)
	if c.Count != 3 {
		t.Errorf("expected Count=3, got %d", c.Count)
	}
}

func TestBoundedCounter_RejectsIncrementBeyondMax(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure")
		}
	}()
	c := &BoundedCounter{MaxCount: 5}
	c.init()
	c.Increment(6)
}

func TestBoundedCounter_Reset(t *testing.T) {
	c := &BoundedCounter{MaxCount: 10}
	c.init()
	c.Increment(7)
	if c.Count != 7 {
		t.Errorf("expected Count=7, got %d", c.Count)
	}
	c.Reset()
	if c.Count != 0 {
		t.Errorf("expected Count=0, got %d", c.Count)
	}
}

func TestBoundedCounter_Compile(t *testing.T) {
	if err := runar.CompileCheck("BoundedCounter.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}

File: examples/rust/property-initializers/BoundedCounter_test.rs
#[path = "BoundedCounter.runar.rs"]
mod contract;

use contract::*;

// NOTE: Contracts with property initializers define an `init()` method that sets
// default values (e.g., count = 0, active = true). In native Rust tests, you must
// call `init()` after constructing the struct to apply these defaults — the Rust
// struct literal requires all fields, but `init()` overwrites them with the
// intended initial values. At compile time, the Rúnar compiler bakes these
// defaults directly into the Bitcoin Script.

#[test]
fn test_default_initializers() {
    let mut c = BoundedCounter { count: 0, max_count: 10, active: false };
    // Call init() to apply property initializer defaults
    c.init();
    assert_eq!(c.count, 0);
    assert_eq!(c.active, true);
    c.increment(1);
    assert_eq!(c.count, 1);
}

#[test]
fn test_increment() {
    let mut c = BoundedCounter { count: 0, max_count: 10, active: false };
    c.init();
    c.increment(3);
    assert_eq!(c.count, 3);
}

#[test]
#[should_panic]
fn test_rejects_increment_beyond_max() {
    let mut c = BoundedCounter { count: 0, max_count: 5, active: false };
    c.init();
    c.increment(6);
}

#[test]
fn test_reset() {
    let mut c = BoundedCounter { count: 0, max_count: 10, active: false };
    c.init();
    c.increment(7);
    assert_eq!(c.count, 7);
    c.reset();
    assert_eq!(c.count, 0);
}

#[test]
fn test_compile() {
    runar::compile_check(include_str!("BoundedCounter.runar.rs"), "BoundedCounter.runar.rs").unwrap();
}
